stdlib/immutablepriorityqueue.gr

```diff
@@ -0,0 +1,332 @@
+ */
+record ImmutablePriorityQueue<a> {
+  comp: (a, a) -> Number,
+  size: Number,
+  root: Option<PQRoot<a>>,
+}
+
+/**
+ * @section Values: Functions for working with ImmutablePriorityQueues.
+ */
+
+/**
+ * Creates a new priority queue with a comparator function, which is used to
+ * determine priority of elements. The comparator function takes two elements
+ * and must return 0 if both share priority, a positive number if the first
+ * has greater priority, and a negative number if the first has less priority.
+ * 
+ * @param comp: The comparator function used to indicate priority order
+ * @returns An empty priority queue
+ * 
+ * @example ImmutablePriorityQueue.make(compare) // creates a min priority queue of numbers using the compare pervasive
+ * @example ImmutablePriorityQueue.make((a, b) => String.length(b) - String.length(a)) // creates a priority queue by string length (longest to shortest)
+ *
+ * @since v0.5.3
+ */
+export let make = comp => {
+  { comp, size: 0, root: None }
+}
+
+/**
+ * Returns the number of elements in the priority queue.
+ * 
+ * @param pq: The priority queue to get the size of
+ * @returns The number of elements in the priority queue
+ *
+ * @since v0.5.3
+ */
+export let size = pq => {
+  pq.size
+}
+
+/**
+ * Returns whether or not the priority queue is empty.
+ * 
+ * @param pq: The priority queue to check
+ * @returns `true` if the priority queue has no elements and `false` otherwise
+ *
+ * @since v0.5.3
+ */
+export let isEmpty = pq => {
+  pq.size == 0
+}
+
+let skewLinkNodes = (comp, newNode, node1, node2) => {
+  // make the two nodes with larger values children of the node with the smallest value
+  if (comp(node1.val, newNode.val) <= 0 && comp(node1.val, node2.val) <= 0) {
+    {
+      val: node1.val,
+      rank: node1.rank + 1,
+      children: [newNode, node2, ...node1.children],
+    }
+  } else if (
+    comp(node2.val, newNode.val) <= 0 && comp(node2.val, node1.val) <= 0
+  ) {
+    {
+      val: node2.val,
+      rank: node2.rank + 1,
+      children: [newNode, node1, ...node2.children],
+    }
+  } else {
+    { val: newNode.val, rank: node1.rank + 1, children: [node1, node2] }
+  }
+}
+
+let skewInsert = (comp, val, pq) => {
+  let newNode = { val, rank: 0, children: [] }
+  match (pq) {
+    // the only time two trees will have the same rank is if they are the
+    // smallest-ranked trees in the queue, in which case we should link
+    // them with the new node
+    [node1, node2, ...rest] when node1.rank == node2.rank =>
+      [skewLinkNodes(comp, newNode, node1, node2), ...rest],
+    _ => [newNode, ...pq],
+  }
+}
+
+/**
+ * Returns a new priority queue with the given element inserted into the given priority queue.
+ * 
+ * @param val: The value to add into the priority queue
+ * @param pq: The priority queue
+ * @returns A new priority queue with the given element inserted into the given priority queue
+ */
+export let push = (val, pq) => {
+  let { comp, size, root } = pq
```

Comment:
I left it as `pq` because I assumed that grain doc uses the parameter name when generating documentation

Reply:
👍  That makes sense. We don't enforce that right now because the identifiers are not in the module interface, but it's something we want to do in the future.